### PR TITLE
[WIP]Use DIG to create uhttp client

### PR DESCRIPTION
**Pull request by alsamylkin**
Simplify client construction and use a DI graph to resolve auth and tracing.
If they are not present, an error is returned.

Constructor now accepts 2 options: di graph and middleware,
which can be used to override auth/tracing and add extra middleware.

TODO: Put tracer and auth in the default graph.

### CODE REVIEW

File: modules/uhttp/client/client.go [hunk 1]
```diff
@@ -25,19 +25,83 @@ import (
...
+	"github.com/opentracing/opentracing-go"
+)
+
+const (
+	_middlewareKey = "httpClientMiddleware"
+	_graphKey      = "httpClientGraph"
 )
 
+// WithOutbound lets you to add a custom middleware to the client
+func WithOutbound(middleware ...OutboundMiddleware) modules.Option {
```
Comment (anuptalwalkar) [inline modules/uhttp/client/client.go]: rename `WithOutboundMiddleware` to keep it consistent with rpc

File: modules/uhttp/client/client.go [hunk 2]
```diff
@@ -25,19 +25,83 @@ import (
...
+	return func(info *service.ModuleCreateInfo) error {
+		items := info.Items
+		if m, ok := items[_middlewareKey]; ok {
+			items[_middlewareKey] = append(m.([]OutboundMiddleware), middleware...)
```
Comment (anuptalwalkar) [inline modules/uhttp/client/client.go]: should check for type assertion?

File: modules/uhttp/client/client.go [hunk 3]
```diff
@@ -25,19 +25,83 @@ import (
...
+	}
+
+	for _, opt := range options {
+		if err := opt(&info); err != nil {
+			return nil, err
+		}
+	}
+
+	// Resolve auth and tracer.
+	graph := dig.DefaultGraph()
```
Comment (anuptalwalkar) [inline modules/uhttp/client/client.go]: I don't like this approach of Resolving dependency in New. If you have 10 vars, this method would linearly stretch.  May be refactor it into a method on the client for resolve. Or, have the variables be Resolved in the outbound filters?

File: modules/uhttp/client/client.go [hunk 4]
```diff
@@ -25,19 +25,83 @@ import (
...
+	}
+
+	var auth *auth.CreateAuthInfo
+	if err := graph.Resolve(&auth); err != nil {
+		return nil, err
+	}
+
+	// Make middleware.
+	middleware := make([]OutboundMiddleware, 0, len(options)+2)
+	middleware = append(middleware, tracingOutbound(*tracer), authenticationOutbound(*auth))
```
Comment (anuptalwalkar) [inline modules/uhttp/client/client.go]: Can this be wrapped into some defeultOutboundMiddleware?

File: service/host.go [hunk 5]
```diff
@@ -164,6 +164,7 @@ func (s *manager) shutdown(err error, reason string, exitCode *int) (bool, error
 	}
 
 	// Flush tracing buffers
+	// GFM-378 uhttp.Client is not safe to send requests anymore
```
Comment (anuptalwalkar) [inline service/host.go]: It this an open task, or why?

File: modules/uhttp/client/client.go [hunk 6]
```diff
@@ -25,19 +25,83 @@ import (
...
 // they are going to be applied in following order: tracing, auth, remaining outbound middleware
 // and only if all of them passed the request is going to be send.
-// Client is safe to use by multiple go routines, if global tracer is not changed.
-func New(info auth.CreateAuthInfo, middleware ...OutboundMiddleware) *http.Client {
-	defaultMiddleware := []OutboundMiddleware{tracingOutbound(), authenticationOutbound(info)}
-	defaultMiddleware = append(defaultMiddleware, middleware...)
+func New(options ...modules.Option) (*http.Client, error) {
+
+	// Apply options.
+	info := service.ModuleCreateInfo{
```
Comment (anuptalwalkar) [inline modules/uhttp/client/client.go]: I am confused about info creation here. Why is it created with an empty map?

File: modules/uhttp/client/client.go (hunk 2 above)
Comment (alsamylkin) [inline modules/uhttp/client/client.go]: I am consistent with current usage, I think we never check these: e.g. inboundMiddlewareFromCreateInfo, because this really means something went really wrong

File: modules/uhttp/client/client.go (hunk 6 above)
Comment (alsamylkin) [inline modules/uhttp/client/client.go]: So we can add elements to it later in options..

File: modules/uhttp/client/client.go (hunk 3 above)
Comment (alsamylkin) [inline modules/uhttp/client/client.go]: If we'll have 10 default(!) arguments, I'll refactor, but just for, it is an overkill

File: service/host.go (hunk 5 above)
Comment (alsamylkin) [inline service/host.go]: I opened it when starting working on moving tracer out, we close a jaeger reporter here and tracer will panic if any requests will be made by http client

Comment (madhuravi) [inline service/host.go]: This is not unique to http. If Stop is called, nothing is safe to call. Doesn't this happen to YARPC client too?

**Comment by alsamylkin** [comment]
Clients for YARPC are created with a dispatcher and we stop it in modules
before, so it should be ok(error I assume).

File: service/host.go (hunk 5 above)
Comment (madhuravi) [inline service/host.go]: From Alex - "Clients for YARPC are created with a dispatcher and we stop it in modules
before, so it should be ok(error I assume)."

Yes but it's not a module. Clients are not necessarily aware that dispatcher comes from initializing the YARPC module. It is fetched separately from module initialization.

File: modules/uhttp/client/client.go [hunk 7]
```diff
@@ -25,19 +25,83 @@ import (
...
 	"go.uber.org/fx/auth"
+	"go.uber.org/fx/dig"
+	"go.uber.org/fx/modules"
+	"go.uber.org/fx/service"
+
+	"github.com/opentracing/opentracing-go"
+)
+
+const (
+	_middlewareKey = "httpClientMiddleware"
```
Comment (madhuravi) [inline modules/uhttp/client/client.go]: If it's outboundMiddleware otherwise, the key should match that.

File: modules/uhttp/client/client.go [hunk 8]
```diff
@@ -25,19 +25,83 @@ import (
...
+			items[_middlewareKey] = append(m.([]OutboundMiddleware), middleware...)
+		} else {
+			items[_middlewareKey] = middleware
+		}
+		return nil
+	}
+}
+
+// WithGraph allows you to use a custom dependency injection graph to resolve Tracer and AuthInfo
+func WithGraph(graph dig.Graph) modules.Option {
```
Comment (madhuravi) [inline modules/uhttp/client/client.go]: Why is this a public function? I think it's too much for clients to think about.
client.New(WithGraph())

They shouldn't have to know that a client is initialized with a graph or set it up etc.

Is our goal to have separate graphs per module? Why doesn't using the global graph work?
Maybe we should figure out the strategy for how to use dig before we start using it everywhere.

File: modules/uhttp/client/client.go [hunk 9]
```diff
@@ -25,19 +25,83 @@ import (
...
+		if err := opt(&info); err != nil {
+			return nil, err
+		}
+	}
+
+	// Resolve auth and tracer.
+	graph := dig.DefaultGraph()
+	if g, ok := info.Items[_graphKey]; ok {
+		graph = g.(dig.Graph)
+	}
```
Comment (madhuravi) [inline modules/uhttp/client/client.go]: why not 
else {
  graph = dig.DefaultGraph()
}
instead of always assigning and then over-writing. Avoids the overhead of DefaultGraph if provided.

File: modules/uhttp/client/client_middleware_test.go [hunk 10]
```diff
@@ -115,38 +115,35 @@ func TestExecutionChainOutboundMiddleware_AuthContextPropagationFailure(t *testi
 }
 
 func TestOutboundMiddlewareWithTracerErrors(t *testing.T) {
+	t.Parallel()
+	tr := &shadowTracer{
+		opentracing.NoopTracer{},
+		func(sm opentracing.SpanContext, format interface{}, carrier interface{}) error {
```
Comment (madhuravi) [inline modules/uhttp/client/client_middleware_test.go]: nit: sm?

File: modules/uhttp/client/client_middleware_test.go [hunk 11]
```diff
@@ -115,38 +115,35 @@ func TestExecutionChainOutboundMiddleware_AuthContextPropagationFailure(t *testi
...
+			return errors.New("Very bad _tracer")
```
Comment (madhuravi) [inline modules/uhttp/client/client_middleware_test.go]: nit: I would skip the underscore. The shadowTracer is not aware of the _tracer.

File: modules/uhttp/client/client_test.go [hunk 12]
```diff
@@ -33,96 +38,134 @@ var (
 	_testYaml = []byte(`
 name: test
 `)
-	_testClient = New(fakeAuthInfo{yaml: _testYaml})
+	_tracer   = opentracing.NoopTracer{}
```
Comment (madhuravi) [inline modules/uhttp/client/client_test.go]: call this noopTracer, makes it more readable

File: service/host.go (hunk 5 above)
Comment (alsamylkin) [inline service/host.go]: If clients get dispatcher with yarpc.Dispatcher(), but don't initialize the module, then it will be nil and users will not be able to create client..

File: modules/uhttp/client/client.go (hunk 8 above)
Comment (alsamylkin) [inline modules/uhttp/client/client.go]: And they will not have to know, a regular `client.New()` will return a client, that is initialized with a default graph.

They may need to override a tracer/auth client, then they can provide their DI graph, that's why it is public. No one forces them to use it.

File: modules/uhttp/client/client.go (hunk 9 above)
Comment (alsamylkin) [inline modules/uhttp/client/client.go]: In most cases we expect people to not provide it, that's why I've initialized graph that way.

File: modules/uhttp/client/client_middleware_test.go (hunk 11 above)
Comment (alsamylkin) [inline modules/uhttp/client/client_middleware_test.go]: This is gogland's `refactoring`.. :(

File: modules/uhttp/client/client.go (hunk 1 above)
Comment (alsamylkin) [inline modules/uhttp/client/client.go]: Sure, it just becomes so much SimpleBeanFactoryAwareAspectInstanceFactory..

File: modules/uhttp/client/client.go (hunk 4 above)
Comment (alsamylkin) [inline modules/uhttp/client/client.go]: Yes, it is not very concise, let me rephrase it..

File: service/host.go (hunk 5 above)
Comment (madhuravi) [inline service/host.go]: yes but they are not usually aware of this at all. They don't *know* that this is how it works and they don't have. UberFx's thing is all this functionality works only if you initialize and start the service.

**Comment by alsamylkin** [comment]
Looks like it is easier to start over and pass all the parameters to the constructor explicitly.
